services/workflows-service/src/main.ts

```diff
@@ -50,6 +51,28 @@ const corsOrigins = [
 ];
 
 const main = async () => {
```

Comment:
in `services/workflows-service/src/auth/auth.service.ts` you can inject a service using `factory` function 
<img width="376" alt="image" src="https://github.com/user-attachments/assets/e54e3958-c28e-4723-a378-8d4d31e3132a">

## Import related service when configuration supabase client 
```ts
@Module({
  imports: [ConfigModule,...],
  providers: [
  ....
    {
      provide: ISupabaseService,
      useFactory: (configService: ConfigService) => {
        const apiKey = configService.get<string>('TELEMETRY_SUPABASE_API_KEY');
        if (apiKey) {
          return new RealSupabaseService(configService);
        }
        return new MockSupabaseService();
      },
      inject: [ConfigService],
    },
    LocalAuthGuard,
  ],
  exports: [ISupabaseService, ...],
})
export class AuthModule {}
```
## 
```ts
// i-supabase-service.ts
export interface ISupabaseService {
  logSignIn(url: string): Promise<void>;
}

// mock-supabase.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ISupabaseService } from './i-supabase-service';

@Injectable()
export class MockSupabaseService implements ISupabaseService {
  private readonly logger = new Logger(MockSupabaseService.name);

  async logSignIn(url: string): Promise<void> {
    this.logger.debug(`Mock log: Sign-in recorded for URL: ${url}`);
  }
}

// real-supabase.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { createClient, SupabaseClient } from '@supabase/supabase-js';
import { ISupabaseService } from './i-supabase-service';

@Injectable()
export class RealSupabaseService implements ISupabaseService {
  private readonly supabaseClient: SupabaseClient;

  constructor(private readonly configService: ConfigService) {
    const supabaseUrl = this.configService.get<string>('SUPABASE_URL');
    const supabaseApiKey = this.configService.get<string>('TELEMETRY_SUPABASE_API_KEY');

    if (!supabaseUrl || !supabaseApiKey) {
      throw new Error('Supabase URL or API key is missing in configuration');
    }

    this.supabaseClient = createClient(supabaseUrl, supabaseApiKey);
    this.logger.log('Supabase client initialized.');
  }

  async logSignIn(url: string): Promise<void> {
    const { data, error } = await this.supabaseClient.from('logins').insert([{ url }]);

    if (error) {
      this.logger.error(`Failed to log sign-in: ${error.message}`);
    } else {
      this.logger.log(`Sign-in logged successfully: ${data}`);
    }
  }
}
```

Reply:
## Suggested Improvements
I suggest refactoring the Supabase client logic out of the guard itself. Currently, this logic is embedded directly within the canActivate method, which violates the single responsibility principle and tightly couples the guard with Supabase functionality. Here’s a suggested approach:

## Extract Supabase Logic into a Service:

Move Supabase-specific logic into a separate service, for example, SupabaseService. This service will handle client initialization and interactions with the Supabase API.

## Service factory when env var defined
Define an interface (ISupabaseService) that both a real and a mock implementation will follow. Use a provider to choose between (Real)SupabaseService (with actual Supabase client) and MockSupabaseService (no-op implementation) based on the presence of environment variables.